packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -0,0 +1,434 @@
+
+export interface EditableTreeContext {
+    /**
+     * Call before editing.
+     *
+     * Note that after performing edits, EditableTrees for nodes that no longer exist are invalid to use.
+     * TODO: maybe add an API to check if a specific EditableTree still exists,
+     * and only make use other than that invalid.
+     */
+    prepareForEdit(): void;
+
+    /**
+     * Call to free resources.
+     * EditableTrees created in this context are invalid to use after this.
+     */
+    free(): void;
+}
+
+class ProxyContext implements EditableTreeContext {
+    public readonly withCursors: Set<ProxyTarget> = new Set();
+    public readonly withAnchors: Set<ProxyTarget> = new Set();
+    constructor(public readonly forest: IEditableForest) {}
+
+    public prepareForEdit(): void {
+        for (const target of this.withCursors) {
+            target.prepareForEdit();
+        }
+        assert(this.withCursors.size === 0, "prepareForEdit should remove all cursors");
+    }
+    public free(): void {
+        for (const target of this.withCursors) {
+            target.free();
+        }
+        for (const target of this.withAnchors) {
+            target.free();
+        }
+        assert(this.withCursors.size === 0, "free should remove all cursors");
+        assert(this.withAnchors.size === 0, "free should remove all anchors");
+    }
+}
+
+class ProxyTarget {
+	private readonly lazyCursor: ITreeSubscriptionCursor;
+    private anchor?: Anchor;
+
+	constructor(
+		public readonly context: ProxyContext,
+		cursor: ITreeSubscriptionCursor,
+	) {
+		this.lazyCursor = cursor.fork();
+        context.withCursors.add(this);
+	}
+
+    public free(): void {
+        this.lazyCursor.free();
+        this.context.withCursors.delete(this);
+        if (this.anchor !== undefined) {
+            this.context.forest.anchors.forget(this.anchor);
+            this.context.withAnchors.delete(this);
+            this.anchor = undefined;
+        }
+    }
+
+    public prepareForEdit(): void {
+        if (this.anchor === undefined) {
+            this.anchor = this.lazyCursor.buildAnchor();
+            this.context.withAnchors.add(this);
+        }
+        this.lazyCursor.clear();
+        this.context.withCursors.delete(this);
+    }
+
+	public get cursor(): ITreeSubscriptionCursor {
+		if (this.lazyCursor.state === ITreeSubscriptionCursorState.Cleared) {
+            assert(this.anchor !== undefined, "EditableTree should have an anchor if it does not have a cursor");
+			const result = this.context.forest.tryMoveCursorTo(this.anchor, this.lazyCursor);
+			assert(result === TreeNavigationResult.Ok,
+                "It is invalid to access an EditableTree node which no longer exists");
+            this.context.withCursors.add(this);
+            this.context.forest.anchors.forget(this.anchor);
+            this.context.withAnchors.delete(this);
+            this.anchor = undefined;
```

Comment:
Is there a way we can add test coverage for this?

Reply:
We should at minimum include a TODO (here or in the test suite) to cover ProxyContext's methods.

Reply:
Yes, it will definitely be covered e.g. in https://github.com/sharptrip/FluidFramework/pull/4. I'll add a TODO for now.